### ISSUE TITLE
Request for Dataset Construction Details

### ISSUE THREAD

**Issue by Xiaolinger-Z** [lmu-plai/blens#1]
Thank you for making your dataset publicly available. I am currently working to reproduce your experiments on additional datasets and have a few questions regarding the dataset construction process.

While I found helpful explanations in Appendix A.5, I would appreciate some clarification on:

  - The exact meaning and role of virtual addresses and placeholders in your dataset
  - Whether the binary identifiers and function identifiers are specific to your dataset or follow some rules

To ensure accurate reproduction, would it be possible for you to:
  - Share the dataset processing scripts used in your work, or
  - Provide more detailed documentation about the dataset construction pipeline?
  
I completely understand if there are constraints preventing full disclosure, but any additional information you could share would be immensely helpful.

**Comment by YunruWang** [comment]
Hi Xiaoling,

Thank you for your interest in our work and your thoughtful questions. We're glad to hear you're trying to reproduce the experiments on additional datasets. 

Regarding your first two questions:

We recommend checking the loadData function in [builder.py](http://builder.py/), and referring to the xflBlensXProjectData/xflBlensXBinaryData under data/embeddings/ to understand the data format (data could be found with the [zenodo link](https://zenodo.org/records/15119877) ).

In general, the data format follows that of DEXTER in XFL. The loadData function retrieves the DEXTER, PalmTree, and CLAP embeddings for a given function. Virtual addresses play an important role in identifying functions, especially for CLAP and PalmTree.

Here is a relevant excerpt from [builder.py](http://builder.py/) (line 115-121):

```python
# in RunExp.py, line 134-147, the keys for clap and palmtree are set True, and DEXTER False
if key == True:
    if name in ["clap"]:
        data[name] = d[(binPath, vaddr)]
    else:
        data[name] = d[(binPath, vaddr, realFunctionName)]
else:
    data[name] = d[fId]
```

From this, you can see that:

- The identifier for CLAP embeddings is `(binPath, vaddr)`
- For PalmTree, it's `(binPath, vaddr, realFunctionName)`
- For Dexter, it is simply `fId`

Both `fId` and `bId` are present only in the DEXTER embeddings and are not used outside the context of loading DEXTER embeddings. The `placeholder` serves as a temporary memory location where the tokenizer will insert the tokenized version of the function name. 

To put more on what you could find in xflBlensXProjectData/xflBlensXBinaryData, the data structure consists of three lists: the first for the training set, the second for the validation set, and the third for the testing set. Each element in these lists is a tuple: (binPath, vaddr, realFunctionName, functionName, tokens, bId, fId) 
- **binPath**: The file path of the binary as it existed on the system during analysis (this should be unique).

- **vaddr**: The virtual address of the function should correspond to virtual addresses obtained by PalmTree and CLAP preprocessing.
- **realFunctionName and functionName**: The actual function name in its original form, whereas **functionName** is a minimally transformed version generated by the XFL engine.
- **tokens**: This field is filled in by the tokenizer; prior to tokenization, we chose to duplicate the virtual address as a placeholder.
- **bId and fId**: Both are identifiers assigned by XFL’s PSQL tables as part of the XFL analysis process. Almost all information in each tuple can be retrieved via a PSQL query to the XFL database after analyzing the binaries.

As for the dataset construction pipeline, we largely reused the original pipelines from XFL, PalmTree, and CLAP with minimal modifications. We’re currently in the process of organizing the related scripts for potential release.

In the meantime, we’d be happy to answer any specific questions or clarify parts of the process.
